src/agent/block_store_services/block_store_fs.js

```diff
@@ -88,6 +92,31 @@ class BlockStoreFs extends BlockStoreBase {
         }
     }
 
+    /**
+     * _check_write_space relies on cached df data from `get_storage_info` to check if 
+     * there is enough space for the write. If for some reason the cached data is unavailable,
+     * we fallback to the base class implementation.
+     * 
+     * `get_storage_info` is called every 60 seconds, assuming ~ 500MB/s write speed, our cached
+     * data can be off by 30GB.
+     * 
+     * NOTE: overrides the base class implementation to use the cached df data
+     * @param {*} data_length 
+     */
+    _check_write_space(data_length) {
+        if (!this.cached_df_data) {
+            // It is probably better to reject write if there is even a possibility of
+            // partial write issues.
+            return super._check_write_space(data_length);
+        }
+
+        const required_space = data_length + (1024 * 1024); // require some spare space
+        const free_space = this.cached_df_data.free;
+        if (free_space < required_space) {
+            throw new RpcError('NO_BLOCK_STORE_SPACE', 'used space exceeded the total capacity of ' +
+                this.usage_limit + ' bytes');
+        }
+    }
```

Comment:
How about we make `_check_write_space` async and separate it from get_storage_info(), so that we can decide when to trigger reloading of the df info based on other conditions, instead of just relying on the get_storage_info() timing, something like this:

configs -
```js
config.BLOCK_STORE_FS_CACHED_DF_MAX_TIME = 30 * 1000; // 30 seconds
config.BLOCK_STORE_FS_CACHED_DF_MIN_TIME = 1 * 1000; // 1 seconds
config.BLOCK_STORE_FS_CACHED_DF_MIN_SPACE = 1 * 1024 * 1024 * 1024; // 1 GB
```

check -
```suggestion
    async _check_write_space(data_length) {
        const now = Date.now();
        if (!this.cached_df_data || 
            (now - this.cached_df_data.time >= config.BLOCK_STORE_FS_CACHED_DF_MAX_TIME) ||
            (now - this.cached_df_data.time >= config.BLOCK_STORE_FS_CACHED_DF_MIN_TIME && 
                this.cached_df_data.free <= config.BLOCK_STORE_FS_CACHED_DF_MIN_SPACE)) {
            const drive = await os_utils.get_drive_of_path(this.root_path);
            this.cached_df_data = { ...drive.storage, time: now };
        }

        ...
    }
```